presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -0,0 +1,283 @@
+        // where rows did not exist in the inner table.  For some aggregate functions, such as count, the result
+        // of an aggregation over a single null row is one or zero rather than null. In order to ensure correct results,
+        // we add a coalesce function with the output of the new outer join and the agggregation performed over a single
+        // null row.
+        private PlanNode coalesceWithNullAggregation(AggregationNode aggregationNode, PlanNode outerJoin)
+        {
+            NullLiteral nullLiteral = new NullLiteral();
+            ImmutableList.Builder<Symbol> nullSymbols = ImmutableList.builder();
+            ImmutableList.Builder<Expression> nullLiterals = ImmutableList.builder();
+            ImmutableMap.Builder<Symbol, Expression> aggregationSourceToValuesMapping = ImmutableMap.builder();
+            for (Symbol sourceSymbol : aggregationNode.getSource().getOutputSymbols()) {
+                nullLiterals.add(nullLiteral);
+                Symbol nullSymbol = symbolAllcoator.newSymbol(nullLiteral, symbolAllcoator.getTypes().get(sourceSymbol));
+                nullSymbols.add(nullSymbol);
+                aggregationSourceToValuesMapping.put(sourceSymbol, nullSymbol.toSymbolReference());
+            }
+            ValuesNode values = new ValuesNode(
+                    idAllocator.getNextId(),
+                    nullSymbols.build(),
+                    ImmutableList.of(nullLiterals.build()));
+
+            ImmutableMap.Builder<Symbol, Symbol> basicAggregationToOverNullMappingBuilder = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregationsOverNullBuilder = ImmutableMap.builder();
+            for (Map.Entry<Symbol, FunctionCall> entry : aggregationNode.getAggregations().entrySet()) {
+                FunctionCall overNullFunction = ExpressionTreeRewriter.rewriteWith(new ExpressionSymbolInliner(aggregationSourceToValuesMapping.build()), entry.getValue());
+                Symbol overNullSymbol = symbolAllcoator.newSymbol(overNullFunction, symbolAllcoator.getTypes().get(entry.getKey()));
+                aggregationsOverNullBuilder.put(overNullSymbol, overNullFunction);
+                basicAggregationToOverNullMappingBuilder.put(entry.getKey(), overNullSymbol);
+            }
+
+            ImmutableMap<Symbol, Symbol> basicAggregationToOverNullMapping = basicAggregationToOverNullMappingBuilder.build();
+            AggregationNode aggregationOverNullNode = new AggregationNode(
+                    idAllocator.getNextId(),
+                    values,
+                    aggregationsOverNullBuilder.build(),
+                    aggregationNode.getFunctions().entrySet().stream()
+                            .collect(toImmutableMap(entry -> basicAggregationToOverNullMapping.get(entry.getKey()), Map.Entry::getValue)),
+                    aggregationNode.getMasks(),
```

Comment:
This just works because masks is empty. If not, this would be incorrect, as masks is supposed to be a mapping of the output symbol associated with an aggregation function to the input that contains the boolean to use as a mask.